rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md

```diff
@@ -0,0 +1,61 @@
+| dnnl_use_scaleshift                         | *Inputs*: src, gamma, beta <br><br> *Outputs*: dst                      | *Inputs*: src, gamma, beta <br><br> *Outputs*: dst, mu, sigma^2                      | *Inputs*: diffdst, src, mu, sigma^2, gamma, beta <br><br> *Outputs*: diffsrc, diffgamma, diffbeta | *Inputs*: diffdst, src, mu, sigma^2, gamma, beta, diffgamma, diffbeta <br><br> *Outputs*: diffsrc <br><br> | *Inputs*: diffdst, src, mu, sigma^2, gamma, beta <br><br> *Outputs*: diffgamma, diffbeta <br><br> |
+| dnnl_use_global_stats & dnnl_use_scaleshift | *Inputs*: src, mu, sigma^2, gamma, beta <br><br> *Outputs*: dst         | *Inputs*: src, mu, sigma^2, gamma, beta <br><br> *Outputs*: dst                      | *Inputs*: diffdst, src, mu, sigma^2, gamma, beta <br><br> *Outputs*: diffsrc, diffgamma, diffbeta | *Inputs*: diffdst, src, mu, sigma^2, gamma, beta, diffgamma, diffbeta <br><br> *Outputs*: diffsrc <br><br> | *Inputs*: diffdst, src, mu, sigma^2, gamma, beta <br><br> *Outputs*: diffgamma, diffbeta <br><br> |
+| `flags` & dnnl_fuse_norm_relu               | *Inputs*: same as with `flags` <br><br> *Outputs*: same as with `flags` | *Inputs*: same as with `flags` <br><br> *Outputs*: same as with `flags`, [Workspace] | *Inputs*: same as with `flags`, [Workspace] <br><br> *Outputs*: same as with `flags`              | *Inputs*: same as with `flags`, [Workspace] <br><br> *Outputs*: same as with `flags` <br><br>              | *Inputs*: same as with `flags`, [Workspace] <br><br> *Outputs*: same as with `flags` <br><br>     |
+
+In case of batch normalization the weights are gamma and beta (scale and 
+shift), so supporting backward_weights without use_scaleshift flag doesn't 
+make sense. The behaviour of the individual cases can be summarized in the 
+table below:
+
+| propagation kind | behaviour                                             |
+|------------------|-------------------------------------------------------|
+| DW ss            | calculate data and weights, save scale and shift      |
+| DW               | calculate data and weights, dont save scale and shift |
+| W ss             | calculate weights save scale and shift                |
+| D ss             | calculate data                                        |
+| D                | calculate data and weights, dont save scale and shift |
+| W                | Not supported                                         |
+
+The implementation may just split the current one into two sections (calculate 
+weights and calculate data) and execute it depending on propagation kind.
+
+### Option 2
+Option two is the same as the first except no support for backward_weights. 
+Only the requested part will be implemented. This will reduce the 
```

Comment:
In this option, how would the requested parts be determined? Is it by what memory objects are passed at runtime?

Reply:
I meant that requested by user - backward_weights. Changed for better readability.

Reply:
It's still not clear to me. _How_ does user indicate what parts are requested? Is it by which memory objects are passed at execution time?

Reply:
It can be passed by descriptor. Like for example here: https://github.com/intel-innersource/libraries.performance.math.onednn/blob/aeebe74d6b87e10be658865f25eacd9486994ae9/examples/primitives/batch_normalization.cpp#L107 User can pass `backward_data` or `backward_weights`.

Reply:
Okay, I really do not understand this option. I thought that "only the requested part will be implemented" meant that user could somehow indicate which part of outputs is desired (ie `diffsrc` or `diffgamma` and `diffbeta`), but without relying on the propagation kind. Is that right? Or am I misinterpreting what is meant here by "user" -- do you mean the specific user that requested this feature, or a generic user of the API?

Either way, it would help to be more precise about what Option 2 entails. Thanks!

Reply:
Yes, I mean specific user which requested this feature. This is just an option first without `backward_weights`. I changed this description a little bit.

Reply:
Okay, I finally got it. Thanks!